clients/drcachesim/tools/filter/record_filter.h

```diff
@@ -139,6 +142,15 @@ class record_filter_t : public record_analysis_tool_t {
     parallel_shard_error(void *shard_data) override;
 
 protected:
+    // For core-sharded we need to remember encodings for an input that were
+    // seen on a different core, as there is no reader_t remembering them for us.
+    struct per_input_t {
+        // There should be no contention on the lock as each input is on
```

Comment:
The lock satisfies race detection tools and gives us correctness guarantees.

The scheduler does guarantee that.  I think you're remembering where the unit tests were only checking certain record types so the results didn't advance at the same rate and so the output strings didn't actually line up in time.  Since then the tests were expanded and that check is now present: https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tests/scheduler_unit_tests.cpp#L879